pkg/sql/catalog/schematelemetry/schematelemetrycontroller/controller.go

```diff
@@ -0,0 +1,289 @@
+	pbtypes "github.com/gogo/protobuf/types"
+	"github.com/robfig/cron/v3"
+)
+
+// SchemaTelemetryScheduleName is the name of the schema telemetry schedule.
+const SchemaTelemetryScheduleName = "sql-schema-telemetry"
+
+// SchemaTelemetryRecurrence is the cron-tab string specifying the recurrence
+// for schema telemetry job.
+var SchemaTelemetryRecurrence = settings.RegisterValidatedStringSetting(
+	settings.TenantWritable,
+	"sql.schema.telemetry.recurrence",
+	"cron-tab recurrence for SQL schema telemetry job",
+	"@daily", /* defaultValue */
+	func(_ *settings.Values, s string) error {
+		if _, err := cron.ParseStandard(s); err != nil {
+			return errors.Wrap(err, "invalid cron expression")
+		}
+		return nil
+	},
+).WithPublic()
+
+// ErrDuplicatedSchedules indicates that there is already a schedule for SQL
+// schema telemetry jobs existing in the system.scheduled_jobs table.
+var ErrDuplicatedSchedules = errors.New("creating multiple schema telemetry schedules is disallowed")
+
+// ErrVersionGate indicates that SQL schema telemetry jobs or schedules are
+// not supported by the current cluster version.
+var ErrVersionGate = errors.New("SQL schema telemetry jobs or schedules not supported by current cluster version")
+
+// Controller implements the SQL Schema telemetry subsystem control plane.
+// This exposes administrative interfaces that can be consumed by other parts
+// of the database (e.g. status server, builtins) to control the behavior of the
+// SQL schema telemetry subsystem.
+type Controller struct {
+	scheduleController
+	jr *jobs.Registry
+}
+
+// scheduleController is distinct from Controller which implements the full
+// control plane, because tenant migrations don't have access to the jobs
+// registry.
+type scheduleController struct {
+	db *kv.DB
+	ie sqlutil.InternalExecutor
+	st *cluster.Settings
+}
+
+// NewController is a constructor for *Controller.
+//
+// This constructor needs to be called in the sql package when creating a new
+// sql.Server. This is the reason why it and the definition of the Controller
+// object live in their own package separate from schematelemetry.
+func NewController(
+	db *kv.DB, ie sqlutil.InternalExecutor, st *cluster.Settings, jr *jobs.Registry,
+) *Controller {
+	return &Controller{
+		scheduleController: scheduleController{
+			db: db,
+			ie: ie,
+			st: st,
+		},
+		jr: jr,
+	}
+}
+
+// CreateSchemaTelemetryJob implements eval.SchemaTelemetryController.
+func (c *Controller) CreateSchemaTelemetryJob(
+	ctx context.Context, createdByName string, createdByID int64,
+) (id int64, _ error) {
+	if !c.st.Version.IsActive(ctx, clusterversion.SQLSchemaTelemetryScheduledJobs) {
+		return 0, ErrVersionGate
+	}
+	if err := c.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		jobID, err := CreateSchemaTelemetryJob(ctx, c.jr, txn, createdByName, createdByID)
+		id = int64(jobID)
+		return err
+	}); err != nil {
+		return 0, err
+	}
+	c.jr.NotifyToAdoptJobs()
```

Comment:
I'm not sure that this does exactly what you're intending given the use of `CreateAdoptableJobWithTxn`. That function constructs a job without any node in the cluster having a lease on it. Nodes won't attempt to acquire leases on unclaimed jobs until their adoption loop runs. This function tells the local registry to find any jobs it has a lease for but is not running. We now have a better version of this which takes a specific ID. Either way, this function or the better version won't work while you aren't pre-assigning the job to the local node.

Maybe that's fine. It's not clear that any of this is latency sensitive. Assuming it is fine, may I ask you remove this call so that future readers don't think it does anything?

Reply:
TIL! Thanks for the explanation. Will do.